proxy/ParentConsistentHash.cc

```diff
@@ -64,23 +67,127 @@ ParentConsistentHash::getPathHash(HttpRequestData *hrdata, ATSHash64 *h)
+
+  // Use over-ride URL from HttpTransact::State's cache_info.parent_selection_url, if present.
+  URL *ps_url = NULL;
+  Debug("parent_select", "ParentConsistentHash::%s() hrdata->cache_info_parent_selection_url = 0x%lx", __func__,
+        (unsigned long int)hrdata->cache_info_parent_selection_url);
+  if (hrdata->cache_info_parent_selection_url) {
+    ps_url = *(hrdata->cache_info_parent_selection_url);
+    Debug("parent_select", "ParentConsistentHash::%s() ps_url = 0x%lx", __func__, (unsigned long int)ps_url);
+    if (ps_url) {
+      tmp = ps_url->string_get_ref(&len);
+      if (tmp && len > 0) {
+        // Print the over-ride URL
+        if (is_debug_tag_set("parent_select")) {
+          slen = (len > 1023) ? 1023 : len;
+          strncpy(buffer, tmp, slen);
+          buffer[slen] = 0;
+          Debug("parent_select", "ParentConsistentHash::%s() Using Over-Ride String='%s'.", __func__, buffer);
+        }
+        h->update(tmp, len);
+        goto done;
+      }
+    }
+  }
 
   // Always hash on '/' because paths returned by ATS are always stripped of it
   h->update("/", 1);
 
   tmp = url->path_get(&len);
-  if (tmp) {
+
+  if (tmp && len > 0) {
+    // Print Original Path
+    if (is_debug_tag_set("parent_select")) {
+      slen = (len > 1023) ? 1023 : len;
+      strncpy(buffer, tmp, slen);
+      buffer[slen] = 0;
+      Debug("parent_select", "ParentConsistentHash::%s() Original Path='%s'.", __func__, buffer);
+    }
+
+    // Process max_dirs directive.
+    if (max_dirs != 0) {
+      // Determine number of directory components in the path.
+      // NOTE: Leading '/' is gone already.
+      for (int x = 0; x < len; x++) {
+        if (tmp[x] == '/')
+          num_dirs++;
+      }
+      // If max_dirs positive , include directory components from the left up to max_dirs.
+      // If max_dirs negative , include directory components from the left up to num_dirs - ( abs(max_dirs) - 1 ).
+      int limit = 0;
+      if (max_dirs > 0)
+        limit = max_dirs;
+      else if (max_dirs < 0) {
+        int md = abs(max_dirs) - 1;
+        if (md < num_dirs)
+          limit = num_dirs - md;
+        else
+          limit = 0;
+      }
+      if (limit > 0) {
+        int x     = 0;
+        int count = 0;
+        for (; x < len; x++) {
+          if (tmp[x] == '/')
+            count++;
+          if (count == limit) {
+            len = x + 1;
+            break;
+          }
+        }
+      } else {
+        len = 0;
+      }
+    }
+
+    // Print Post max_dirs Path
+    if (is_debug_tag_set("parent_select")) {
+      slen = (len > 1023) ? 1023 : len;
+      strncpy(buffer, tmp, slen);
+      buffer[slen] = 0;
+      Debug("parent_select", "ParentConsistentHash::%s() Post-MD Path='%s'.", __func__, buffer);
+    }
+
+    if (ignore_fname || max_dirs != 0) {
+      int x = len - 1;
+      for (; x >= 0; x--) {
+        char c = tmp[x];
+        if (c == '/')
+          break;
+      }
+      len = x + 1;
```

Comment:
@jpeach -- It is looking for the last occurrence of '/' similar to strrchr(), but that function is not used since this is not a NULL terminated string. I hope this (and the comments and re-factoring requested by @PSUdaemon ) address your concerns.